### ISSUE TITLE
[tsserver] documentation "about how to implement tsClient?"

### ISSUE THREAD

**Issue by angelozerr** [Microsoft/TypeScript#5923]
I'm implementing tsClient inside Eclipse, but it's very hard for me, because there are no documentation about protocol. So I study implementation with sublime, vim and vscode, but it's not my language, so it's a little hard.

So, it should be very cool if you could provide a documentation to implement tsClient, I mean request/response protocol. 

We could take a simple sample: 

 * open a document. 
 * do a completion.
 * change the editor content to show how to synchronize editor with tsserver. There are 2 means to do that:
    * use "change" (I must study why it doesn't work for me, but if I have a documentation, it will help me a lot).
    *  use "tmpfile"
 * do a completion again with the editor which have changed.
 * close the document.

To explain that, you could use starts tsserevr with Windows cmd, and shows request/response. So any user will able to play with your tsserver.

After that you could speak about advanced topic like validation with event.

Thanks for your help!

**Comment by angelozerr** [comment]
@mhegazy have you the intention to write a wiki page about tsserver. I have noticed that 2.0.6 will provide a lot of command like outliningSpans, applyChangedToOpenFiles, implementation, compileOnSaveEmitFile, completion-full, etc

It should be really cool to explain what is the goal of each commands. In my case I would like to consume it inside typescript.java (Eclipse).

Thanks for your help!

**Comment by mhegazy** [comment]
I do have it on my todo list., but do not know when to be frank. We keep the message format and documentation updated in https://github.com/Microsoft/TypeScript/blob/master/src/server/protocol.d.ts. I would use this as the guide. please let us know if there are issues with this file, or that you need more documentation for a specific request/response.

**Comment by mhegazy** [comment]
I have created an initial document about tsserver in https://github.com/Microsoft/TypeScript/wiki/Standalone-Server-(tsserver). Please let me know what specific information you would like to see added.

**Comment by angelozerr** [comment]
Thank a lot @mhegazy!

At first could you add too https://github.com/angelozerr/typescript.java which provides a Java API to consumes tsserver https://github.com/angelozerr/typescript.java/tree/master/core/ts.core that I use for Eclipse but it could be used for other IDE like Netbeans, WebStorm, etc

When I have started to implement a client for tsserver it was difficult for me to understand how to use it and which commands to call. I think it should be very cool if we take a basic sample (like you have started) and set screenshots by adding some information.

For instance if you wish to implement completion, you need to call:

 - open command
 - completion command and add information about filter. Indeed tsserver returns the list of methods of string if you have:

```
let s = "";
s.ch // open completion returns the whole of method of string (and not the methods of string which starts by ch
```

And set a screenshot of completion (VSCode for instance). For instance if you wish to implement Hover, you need to call quickinfo and set a screenshot.

I think too it's important to speak about position. Before 2.06, only line/offset is supported. Since 2.0.6 you can use position:

```
private getPosition(args: protocol.FileLocationRequestArgs, scriptInfo: ScriptInfo): number {
            return args.position !== undefined ? args.position : scriptInfo.lineOffsetToPosition(args.line, args.offset);
        }
```

In my case the doc I'm waiting for is about implementing compile on save. It seems that tsserver 2.0.6 provides new commands (like compileOnSaveEmitFile. How to consumes thoses commands to implement compile on save?

I have noticed too that there are several methods which ends with full. What i sthe goal of thoses methods?

It exists too getCodeFixes, what is the goal of this command? To implement quick fixes perhaps?

In other words, describe each commands and give a usecase.

At the end, how to add custom commands?

I know it's a big work, but I would like just to tell you about my needs.

Anyway thanks a lot again!

**Comment by robatwilliams** [comment]
Couple of notes from my hacking around in case it's useful for someone.

I started with https://github.com/microsoft/TypeScript/wiki/Standalone-Serve, and observing VSCode's `tsserver.log` output while I performed certain actions in its UI to figure out calls I needed to make for similar functionality.

* `open` any file in the project to trigger project loading. This request has no response. You don't need to wait for the `projectLoadingFinish` event before sending requests.
* Send a `seq` on requests, correlate responses using their `req_seq`
* Messages received can be multipart (contain multiple responses/events)

I wonder if there's scope for a `tsserver-client` node library which would provide a promisified RPC interface.
